src/Tecan/Rack/BaseRack.php

```diff
@@ -0,0 +1,82 @@
+<?php declare(strict_types=1);
+
+namespace MLL\Utils\Tecan\Rack;
+
+use Illuminate\Support\Collection;
+
+abstract class BaseRack implements Rack
+{
+    public const EMPTY_POSITION = null;
+
+    /** @var Collection<int, mixed> */
+    public Collection $positions;
+
+    public function __construct()
+    {
+        $this->positions = Collection::times($this->positionCount(), fn () => self::EMPTY_POSITION)
+            ->mapWithKeys(fn ($content, int $position): array => [$position + 1 => $content]);
+    }
+
+    public function id(): ?string
+    {
+        return null;
+    }
+
+    public function toString(): string
+    {
+        return implode(';', [
+            $this->name(),
+            $this->id(),
+            $this->type(),
+        ]);
+    }
+
+    public function assignFirstEmptyPosition(string $name): int
+    {
+        return $this->assignPosition($name, $this->findFirstEmptyPosition());
+    }
+
+    public function assignLastEmptyPosition(string $name): int
+    {
+        return $this->assignPosition($name, $this->findLastEmptyPosition());
+    }
+
+    public function findFirstEmptyPosition(): int
+    {
+        $firstEmpty = $this->positions
+            ->filter(fn ($content): bool => $content === self::EMPTY_POSITION)
+            ->keys()
+            ->first();
+
+        if ($firstEmpty === null) {
+            throw new NoEmptyPositionOnRack();
+        }
+
+        return $firstEmpty;
+    }
+
+    public function findLastEmptyPosition(): int
+    {
+        $lastEmpty = $this->positions
+            ->filter(fn ($content): bool => $content === self::EMPTY_POSITION)
+            ->keys()
+            ->last();
+
+        if ($lastEmpty === null) {
+            throw new NoEmptyPositionOnRack();
+        }
+
+        return $lastEmpty;
+    }
+
+    public function assignPosition(string $name, int $position): int
+    {
+        if (! $this->positions->has($position)) {
+            throw new InvalidPositionOnRack($position, $this);
+        }
+
+        $this->positions[$position] = $name;
+
+        return $position;
+    }
```

Comment:
In anderen Funktionen ist der Wert der Collection vom Typ `mixed` und wird `$content` genannt, hier gehen nun nur `string` rein und heißen `$name`. Was ist richtig?

Reply:
stimmt, das ist gut. Leider konnte ich jetzt mit mixed den php-doc nicht einfügen und habe vorerst erstmal `/** @param string|int|object|null $content */` verwendent.
Wir würde man es mit einem `mixed` machen?

Reply:
Damit der PHPDoc mit `mixed` bleibt muss ein Kommentar dazu, sonst sieht php-cs-fixer die Angabe als redundant - aber PHPStan beschwert sich dass sie fehlt.